wandb/data_types.py

```diff
@@ -411,12 +487,121 @@ def iterrows(self):
         Yields
         ------
         index : int
-            The index of the row.
+            The index of the row. Using this value in other WandB tables
+            will automatically build a relationship between the tables
         row : List[any]
             The data of the row
         """
         for ndx in range(len(self.data)):
-            yield ndx, self.data[ndx]
+            index = _TableIndex(ndx)
+            index.set_table(self)
+            yield index, self.data[ndx]
+
+    def set_pk(self, col_name):
+        # TODO: Docs
+        assert col_name in self.columns
+        self.cast(col_name, _TablePrimaryKeyType())
+
+    def set_fk(self, col_name, table, table_col):
+        # TODO: Docs
+        assert col_name in self.columns
+        assert col_name != self._pk_col
+        self.cast(col_name, _TableForeignKeyType(table, table_col))
+
+    def _update_keys(self, force_last=False):
+        """Updates the known key-like columns based on the current
+        column types. If the state has been updated since
+        the last update, we wrap the data appropriately in the Key classes
+
+        Arguments:
+        force_last: (bool) Determines wrapping the last column of data even if
+        there are no key updates.
+        """
+        _pk_col = None
+        _fk_cols = set()
+
+        # Buildup the known keys from column types
+        c_types = self._column_types.params["type_map"]
+        for t in c_types:
+            if isinstance(c_types[t], _TablePrimaryKeyType):
+                _pk_col = t
+            elif isinstance(c_types[t], _TableForeignKeyType) or isinstance(
+                c_types[t], _TableForeignIndexType
+            ):
+                _fk_cols.add(t)
+
+        # If there are updates to perform, safely update them
+        has_update = _pk_col != self._pk_col or _fk_cols != self._fk_cols
+        if has_update:
+            # If we removed the PK
+            if _pk_col is None and self._pk_col is not None:
+                raise AssertionError(
+                    "Cannot unset primary key (column {})".format(self._pk_col)
+                )
+            # If there is a removed FK
+            if len(self._fk_cols - _fk_cols) > 0:
+                raise AssertionError(
+                    "Cannot unset foreign key. Attempted to unset ({})".format(
+                        self._fk_cols - _fk_cols
+                    )
+                )
+
+            self._pk_col = _pk_col
+            self._fk_cols = _fk_cols
+
+        # Apply updates to data only if there are update or the caller
+        # requested the final row to be updated
+        if has_update or force_last:
+            self._apply_key_updates(not has_update)
+
+    def _apply_key_updates(self, only_last=False):
```

Comment:
Style / design: a boolean switch argument that completely changes what a method does can usually be written better as two separate methods.